src/ui/public/test_harness/test_sharding/root_describe_call_filter.js

```diff
@@ -0,0 +1,81 @@
+/**
+ *  Intercept all calls to mocha.describe() and determine
+ *  which calls make it through using a filter function.
+ *
+ *  The filter function is also only called for ROOT describe()
+ *  calls, and all describe calls inside any allowed call are
+ *  automatically passed through.
+ *
+ *  ## example
+ *
+ *  assume tests that look like this:
+ *
+ *  ```js
+ *  describe('section 1', () => {
+ *    describe('item 1', () => {
+ *
+ *    })
+ *  })
+ *  ```
+ *
+ *  If the filter function returned true for "section 1" then "item 1"
+ *  would automatically be authorized. If it returned false for "section 1"
+ *  then "section 1" would be ignored and "item 1" would never be defined
+ *
+ *  @param {function} test - a function that takes the first argument
+ *                             passed to describe, the sections name
+ *  @return {undefined}
+ */
+export function setupRootDescribeCallFilter(test) {
+  const originalDescribe = window.describe;
+
+  if (!originalDescribe) {
+    throw new TypeError('window.describe must be defined by mocha before test sharding can be setup');
+  }
+
+  /**
+   *  In order to track which describe calls are nested
+   *  within other calls, we track our current level as a
+   *  number. Before passing any describe call to mocha we
+   *  increment that counter, and once mocha is done we
+   *  decrement it. That way, when we are at the root level
+   *  our "depth" will equal 0
+   *
+   *  ```js
+   *  // depth = 0
+   *  describe('section 1', () => {
+   *    // depth = 1
+   *    describe('item 1', () => {
+   *      // depth = 2
+   *    })
+   *    // depth = 1
+   *  })
+   *  // depth = 0
+   *  ```
+   *
+   *  @type {Number}
+   */
+  let describeCallDepth = 0;
+  const ignoredCalls = [];
+
+  // ensure that window.describe isn't messed with by other code,
+  // if it becomes necessary in the future we can make this fancier
+  Object.defineProperty(window, 'describe', {
+    configurable: false,
+    enumerable: true,
+    value: function describeInterceptor(describeName, describeBody) {
+      const context = this;
+
+      const isRootCall = describeCallDepth === 0;
+      const shouldIgnore = isRootCall && Boolean(test(describeName)) === false;
+      if (shouldIgnore) return;
+
+      try {
+        describeCallDepth += 1;
+        originalDescribe.call(context, describeName, describeBody);
+      } finally {
+        describeCallDepth -= 1;
```

Comment:
```javascript
// The nested describe calls have completed, so we can decrement our counter. It's inside of a `finally` block
// so that our counter will reset to 0, even if there are errors in our tests.
```